Flaky NavigationViewOrientationTest

Re-edited:

Original issue is still reproducible 👀 https://github.com/mapbox/mapbox-navigation-android/issues/1209#issuecomment-420012071

~~✅ Addressed in https://github.com/mapbox/mapbox-navigation-android/pull/1247~~

Found 👇 during CI Instrumentation Tests

```
com.mapbox.mapboxsdk.style.sources.CannotAddSourceException: Source mapbox-location-source already exists
at com.mapbox.mapboxsdk.maps.NativeMapView.nativeAddSource(Native Method)
at com.mapbox.mapboxsdk.maps.NativeMapView.addSource(NativeMapView.java:772)
at com.mapbox.mapboxsdk.maps.MapboxMap.addSource(MapboxMap.java:419)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayer.addLocationSource(LocationLayer.java:323)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayer.initializeComponents(LocationLayer.java:108)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayer.<init>(LocationLayer.java:103)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin.initialize(LocationLayerPlugin.java:687)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin.<init>(LocationLayerPlugin.java:184)
at com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin.<init>(LocationLayerPlugin.java:164)
at com.mapbox.services.android.navigation.ui.v5.map.NavigationMapboxMap.initializeLocationLayer(NavigationMapboxMap.java:179)
at com.mapbox.services.android.navigation.ui.v5.map.NavigationMapboxMap.<init>(NavigationMapboxMap.java:71)
at com.mapbox.services.android.navigation.ui.v5.NavigationView.initializeNavigationMap(NavigationView.java:258)
at com.mapbox.services.android.navigation.ui.v5.NavigationView.onMapReady(NavigationView.java:202)
at com.mapbox.mapboxsdk.maps.MapView.getMapAsync(MapView.java:665)
at com.mapbox.services.android.navigation.ui.v5.NavigationView.initialize(NavigationView.java:376)
at testapp.utils.OnNavigationReadyIdlingResource.onResponse(OnNavigationReadyIdlingResource.java:74)
at com.mapbox.api.directions.v5.MapboxDirections$1.onResponse(MapboxDirections.java:169)
at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:70)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.support.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
at android.support.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:462)
at android.support.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:384)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:268)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:265)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6938)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
```

It seems that we're suffering from an occasional race condition in `OnNavigationReadyIdlingResource` when fetching the route and initializing `NavigationView` in `TestNavigationActivity` and `navigationView.initialize` is called twice somehow causing the `CannotAddSourceException` crash.

Noting that I saw the error coming from both tests in `NavigationViewOrientationTest` (`onOrientationLandscape_navigationContinuesRunning` and `onOrientationPortrait_navigationContinuesRunning`) and only in Firebase Test Lab.

For context, these tests were introduced in https://github.com/mapbox/mapbox-navigation-android/pull/1179

Run into the following `SocketTimeoutException` during CI Instrumentation Tests - thrown because there was a failure when requesting Mapbox Directions API https://github.com/mapbox/mapbox-navigation-android/blob/0854e8e0c17113eef980fafe1dbc8a290f5dd35f/app/src/androidTest/java/testapp/utils/OnNavigationReadyIdlingResource.java#L79 

```
java.lang.RuntimeException: java.net.SocketTimeoutException: timeout
at testapp.utils.OnNavigationReadyIdlingResource.onFailure(OnNavigationReadyIdlingResource.java:79)
at com.mapbox.api.directions.v5.MapboxDirections$1.onFailure(MapboxDirections.java:176)
at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$2.run(ExecutorCallAdapterFactory.java:79)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.support.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
at android.support.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:462)
at android.support.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:384)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:268)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:265)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6541)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
Caused by: java.net.SocketTimeoutException: timeout
at okio.Okio$4.newTimeoutException(Okio.java:232)
at okio.AsyncTimeout.exit(AsyncTimeout.java:275)
at okio.AsyncTimeout$2.read(AsyncTimeout.java:243)
at okio.RealBufferedSource.indexOf(RealBufferedSource.java:355)
at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:227)
at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215)
at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
at java.lang.Thread.run(Thread.java:764)
```

Run into a new crash when requesting Mapbox Directions API during CI Instrumentation Tests 
```
java.lang.RuntimeException: javax.net.ssl.SSLHandshakeException: Handshake failed
at testapp.utils.OnNavigationReadyIdlingResource.onFailure(OnNavigationReadyIdlingResource.java:79)
at com.mapbox.api.directions.v5.MapboxDirections$1.onFailure(MapboxDirections.java:176)
at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$2.run(ExecutorCallAdapterFactory.java:79)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.support.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
at android.support.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:462)
at android.support.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:384)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:268)
at android.support.test.espresso.ViewInteraction$2.call(ViewInteraction.java:265)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:789)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6541)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
Caused by: javax.net.ssl.SSLHandshakeException: Handshake failed
at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:423)
at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:318)
at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:282)
at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167)
at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
at java.lang.Thread.run(Thread.java:764)
Caused by: javax.net.ssl.SSLProtocolException: SSL handshake aborted: ssl=0x780dcf1200: Failure in SSL library, usually a protocol error
error:100000df:SSL routines:OPENSSL_internal:UNEXPECTED_MESSAGE (external/boringssl/src/ssl/handshake_client.c:1185 0x781a5b464b:0x00000000)
at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:351)
... 23 more
```

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java
@@ -74,7 +74,7 @@ public class NavigationView extends CoordinatorLayout implements LifecycleObserv
   private NavigationMapboxMap navigationMap;
   private OnNavigationReadyCallback onNavigationReadyCallback;
   private MapboxMap.OnMoveListener onMoveListener;
-  private boolean isInitialized;
+  private boolean isMapInitialized;
 
   public NavigationView(Context context) {
     this(context, null);
@@ -202,6 +202,7 @@ public class NavigationView extends CoordinatorLayout implements LifecycleObserv
   public void onMapReady(MapboxMap mapboxMap) {
     initializeNavigationMap(mapView, mapboxMap);
     onNavigationReadyCallback.onNavigationReady(navigationViewModel.isRunning());
+    isMapInitialized = true;
   }
 
   @Override
@@ -347,11 +348,7 @@ public class NavigationView extends CoordinatorLayout implements LifecycleObserv
    * @param options with containing route / coordinate data
    */
   public void startNavigation(NavigationViewOptions options) {
-    if (!isInitialized) {
-      initializeNavigation(options);
-    } else {
-      navigationViewModel.updateNavigation(options);
-    }
+    initializeNavigation(options);
   }
 
   /**
@@ -374,7 +371,11 @@ public class NavigationView extends CoordinatorLayout implements LifecycleObserv
    */
   public void initialize(OnNavigationReadyCallback onNavigationReadyCallback) {
     this.onNavigationReadyCallback = onNavigationReadyCallback;
-    mapView.getMapAsync(this);
+    if (!isMapInitialized) {
+      mapView.getMapAsync(this);
+    } else {
+      onNavigationReadyCallback.onNavigationReady(navigationViewModel.isRunning());
+    }
   }
 
   /**
@@ -513,7 +514,6 @@ public class NavigationView extends CoordinatorLayout implements LifecycleObserv
     setupNavigationMapboxMap(options);
 
     subscribeViewModels();
-    isInitialized = true;
   }
 
   private void initializeClickListeners() {

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java
@@ -104,6 +104,7 @@ public class NavigationViewModel extends AndroidViewModel {
       locationEngineConductor.onDestroy();
       deactivateInstructionPlayer();
       endNavigation();
+      isRunning = false;
     }
     clearDynamicCameraMap();
     navigationViewEventDispatcher = null;
@@ -188,10 +189,6 @@ public class NavigationViewModel extends AndroidViewModel {
       navigationViewRouteEngine.extractRouteOptions(options);
     }
     return navigation;
-  }
-
-  void updateNavigation(NavigationViewOptions options) {
-    navigationViewRouteEngine.extractRouteOptions(options);
   }
 
   void updateFeedbackScreenshot(String screenshot) {

